src/order/filter/bases.lean

```diff
@@ -209,6 +393,257 @@ lemma has_basis.prod (hla : la.has_basis pa sa) (hlb : lb.has_basis pb sb) :
+end
+
+lemma has_antimono_basis {f : filter α} (h : f.is_countably_generated) :
+ ∃ x : ℕ → set α, f.has_antimono_basis (λ _, true) x :=
+begin
+  rcases h.exists_antimono_seq with ⟨x, x_dec, rfl⟩,
+  use x,
+  constructor,
+  apply has_basis_infi_principal,
+  apply directed_of_mono, apply x_dec,
+  use 0,
+  simpa using x_dec,
+  exact monotone_const
+end
+
+end is_countably_generated
+
+lemma is_countably_generated_seq (x : ℕ → set α) : is_countably_generated (⨅ i, principal $ x i) :=
+begin
+  rcases antimono_seq_of_seq x with ⟨y, am, h⟩,
+  rw h,
+  use [range y, countable_range _],
+  rw (has_basis_infi_principal _ _).eq_generate,
+  { simp [range] },
+  { apply directed_of_mono, apply am },
+  { use 0 },
+end
+
+lemma is_countably_generated_of_seq {f : filter α} (h : ∃ x : ℕ → set α, f = ⨅ i, principal $ x i) :
+  f.is_countably_generated  :=
+let ⟨x, h⟩ := h in by rw h ; apply is_countably_generated_seq
+
+lemma is_countably_generated_binfi_principal {B : set $ set α} (h : countable B) :
+  is_countably_generated (⨅ (s ∈ B), principal s) :=
+is_countably_generated_of_seq (countable_binfi_principal_eq_seq_infi h)
+
+lemma is_countable_generated_iff_exists_antimono_basis {f : filter α} : is_countably_generated f ↔
+  ∃ x : ℕ → set α, f.has_antimono_basis (λ _, true) x :=
+begin
+  split,
+  { intro h,
+    exact h.has_antimono_basis },
+  { rintros ⟨x, h⟩,
+    rw h.to_has_basis.eq_infi,
+    exact is_countably_generated_seq x },
+end
+
+namespace is_countably_generated
+
+lemma exists_antimono_seq' {f : filter α} (cblb : f.is_countably_generated) :
+  ∃ x : ℕ → set α, (∀ i j, i ≤ j → x j ⊆ x i) ∧ ∀ {s}, (s ∈ f ↔ ∃ i, x i ⊆ s) :=
+let ⟨x, hx⟩ := is_countable_generated_iff_exists_antimono_basis.mp cblb in
+⟨x, λ i j, hx.decreasing trivial trivial, λ s, by simp [hx.to_has_basis.mem_iff]⟩
+
+protected lemma comap {l : filter β} (h : l.is_countably_generated) (f : α → β) :
+  (comap f l).is_countably_generated :=
+begin
+  rcases h.exists_seq with ⟨x, hx⟩,
+  apply is_countably_generated_of_seq,
+  use λ i, f ⁻¹' x i,
+  calc
+    comap f l = comap f (⨅ i, principal (x i))   : by rw hx
+          ... = (⨅ i, comap f $ principal $ x i) : comap_infi
+          ... = (⨅ i, principal $ f ⁻¹' x i)     : by simp_rw comap_principal,
+end
+
+/-- An abstract version of continuity of sequentially continuous functions on metric spaces:
+if a filter `k` is countably generated then `tendsto f k l` iff for every sequence `u`
+converging to `k`, `f ∘ u` tends to `l`. -/
+lemma tendsto_iff_seq_tendsto {f : α → β} {k : filter α} {l : filter β}
+  (hcb : k.is_countably_generated) :
+  tendsto f k l ↔ (∀ x : ℕ → α, tendsto x at_top k → tendsto (f ∘ x) at_top l) :=
+suffices (∀ x : ℕ → α, tendsto x at_top k → tendsto (f ∘ x) at_top l) → tendsto f k l,
+  from ⟨by intros; apply tendsto.comp; assumption, by assumption⟩,
+begin
+  rcases hcb.exists_antimono_seq with ⟨g, gmon, gbasis⟩,
+  have gbasis : ∀ A, A ∈ k ↔ ∃ i, g i ⊆ A,
```

Comment:
How come this is not one of the lemmas that are already available on bases, and that you need to prove it again here?

Reply:
I didn't write this lemma, it was moved from `topology.bases` (and maybe adjusted slightly to the new setup, I don't remember). But I'll think about proving a more general version later (unless you do it first and push this branch).

Reply:
Don't worry, it can wait for a later PR.